trio/_file_io.py

```diff
@@ -48,59 +74,199 @@
+if TYPE_CHECKING:
+    from typing_extensions import Buffer, Protocol
+
+    # fmt: off
+
+    class _HasClosed(Protocol):
+        @property
+        def closed(self) -> bool: ...
+
+    class _HasEncoding(Protocol):
+        @property
+        def encoding(self) -> str: ...
+
+    class _HasErrors(Protocol):
+        @property
+        def errors(self) -> str | None: ...
+
+    class _HasFileNo(Protocol):
+        def fileno(self) -> int: ...
+
+    class _HasIsATTY(Protocol):
+        def isatty(self) -> bool: ...
+
+    class _HasNewlines(Protocol):
+        @property
+        def newlines(self) -> Any: ...  # None, str or tuple
+
+    class _HasReadable(Protocol):
+        def readable(self) -> bool: ...
+
+    class _HasSeekable(Protocol):
+        def seekable(self) -> bool: ...
+
+    class _HasWritable(Protocol):
+        def writable(self) -> bool: ...
+
+    class _HasBuffer(Protocol):
+        @property
+        def buffer(self) -> BinaryIO: ...
+
+    class _HasRaw(Protocol):
+        @property
+        def raw(self) -> io.RawIOBase: ...
+
+    class _HasLineBuffering(Protocol):
+        @property
+        def line_buffering(self) -> bool: ...
+
+    class _HasCloseFD(Protocol):
+        @property
+        def closefd(self) -> bool: ...
+
+    class _HasName(Protocol):
+        @property
+        def name(self) -> str: ...
+
+    class _HasMode(Protocol):
+        @property
+        def mode(self) -> str: ...
+
+    class _CanGetValue(Protocol[AnyStr_co]):
+        def getvalue(self) -> AnyStr_co: ...
+
+    class _CanGetBuffer(Protocol):
+        def getbuffer(self) -> memoryview: ...
+
+    class _CanFlush(Protocol):
+        def flush(self) -> None: ...
+
+    class _CanRead(Protocol[AnyStr_co]):
+        def read(self, __size: int | None = ...) -> AnyStr_co: ...
+
+    class _CanRead1(Protocol):
+        def read1(self, __size: int | None = ...) -> bytes: ...
+
+    class _CanReadAll(Protocol[AnyStr_co]):
+        def readall(self) -> AnyStr_co: ...
+
+    class _CanReadInto(Protocol):
+        def readinto(self, __buf: Buffer) -> int | None: ...
+
+    class _CanReadInto1(Protocol):
+        def readinto1(self, __buffer: Buffer) -> int: ...
+
+    class _CanReadLine(Protocol[AnyStr_co]):
+        def readline(self, __size: int = ...) -> AnyStr_co: ...
+
+    class _CanReadLines(Protocol[AnyStr]):
+        def readlines(self, __hint: int = ...) -> list[AnyStr]: ...
+
+    class _CanSeek(Protocol):
+        def seek(self, __target: int, __whence: int = 0) -> int: ...
+
+    class _CanTell(Protocol):
+        def tell(self) -> int: ...
+
+    class _CanTruncate(Protocol):
+        def truncate(self, __size: int | None = ...) -> int: ...
+
+    class _CanWrite(Protocol[AnyStr_contra]):
+        def write(self, __data: AnyStr_contra) -> int: ...
+
+    class _CanWriteLines(Protocol[T_contra]):
+        """The lines parameter varies for bytes/str, so use a typevar to make the async match."""
```

Comment:
This is probably supposed to be a comment?

It would be nice if `T_contra` was specialized to be `bound=bytes | str` but that's enforced anyways. I'm fine without that change.